urlc5fdb179cc_lagodiuk_evo_neural_network_agents_tgz-pJ8-com_lagodiuk_agent_evolution_MainJ8/src/com/lagodiuk/agent/evolution/Main.java

```diff
@@ -0,0 +1,604 @@
+		dynamicFoodRadioButton = new JRadioButton("dynamic food");
+		foodTypeButtonGroup = new ButtonGroup();
+		foodTypeButtonGroup.add(staticFoodRadioButton);
+		foodTypeButtonGroup.add(dynamicFoodRadioButton);
+		controlsPanel.add(staticFoodRadioButton);
+		controlsPanel.add(dynamicFoodRadioButton);
+		if (staticFood) {
+			staticFoodRadioButton.setSelected(true);
+		} else {
+			dynamicFoodRadioButton.setSelected(true);
+		}
+
+		regenerateFoodCheckbox = new JCheckBox("regenerate food");
+		regenerateFoodCheckbox.setSelected(regenerateFood);
+		controlsPanel.add(regenerateFoodCheckbox);
+
+		playPauseButton = new JButton("pause");
+		controlsPanel.add(playPauseButton);
+
+		resetButton = new JButton("reset");
+		controlsPanel.add(resetButton);
+
+		progressBar = new JProgressBar(0, 100);
+		progressBar.setValue(0);
+		progressBar.setVisible(false);
+		appFrame.add(progressBar, BorderLayout.SOUTH);
+
+		populationInfoLabel = new JLabel("Population: " + populationNumber, SwingConstants.CENTER);
+		appFrame.add(populationInfoLabel, BorderLayout.NORTH);
+
+		prefs = Preferences.userNodeForPackage(Main.class);
+		String brainsDirPath = prefs.get(PREFS_KEY_BRAINS_DIRECTORY, "");
+		fileChooser = new JFileChooser(new File(brainsDirPath));
+	}
+
+	protected static void initializeChangingFoodTypeFunctionality() {
+		ItemListener changingFoodTypeListener = new ItemListener() {
+			@Override
+			public void itemStateChanged(ItemEvent e) {
+				if (e.getStateChange() == ItemEvent.SELECTED) {
+					disableControls();
+					boolean wasPlaying = play;
+					play = false;
+
+					staticFood = !staticFood;
+
+					List<Food> food = new LinkedList<Food>();
+					for (Food f : environment.filter(Food.class)) {
+						food.add(f);
+					}
+					for (Food f : food) {
+						environment.removeAgent(f);
+
+						Food newFood = createRandomFood(1, 1);
+						newFood.setX(f.getX());
+						newFood.setY(f.getY());
+
+						environment.addAgent(newFood);
+					}
+
+					play = wasPlaying;
+					enableControls();
+				}
+			}
+		};
+		staticFoodRadioButton.addItemListener(changingFoodTypeListener);
+		dynamicFoodRadioButton.addItemListener(changingFoodTypeListener);
+	}
+
+	public static void initializeRegenerateFoodCheckboxFunctionality() {
+		regenerateFoodCheckbox.addActionListener(new ActionListener() {
+			@Override
+			public void actionPerformed(ActionEvent arg0) {
+				regenerateFood = !regenerateFood;
+			}
+		});
+	}
+
+	private static void initializeLoadBrainButtonFunctionality() {
+		loadBrainButton.addActionListener(new ActionListener() {
+			@Override
+			public void actionPerformed(ActionEvent arg0) {
+				disableControls();
+
+				int returnVal = fileChooser.showOpenDialog(appFrame);
+				if (returnVal == JFileChooser.APPROVE_OPTION) {
+					try {
+						File brainFile = fileChooser.getSelectedFile();
+						prefs.put(PREFS_KEY_BRAINS_DIRECTORY, brainFile.getParent());
+
+						FileInputStream in = new FileInputStream(brainFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at line 341. The resource is closed at line 342. There are other execution paths that do not contain closure statements, for example, when NeuralNetwork.unmarsall throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).